// plat-import-lib/src/_nav.js
import concat from 'lodash/concat'
const DUMMY_CLIENT_ID = process.env.VUE_APP_PF_CLIENT_ID || ''

export default {
  name: 'pf',
  clientId: DUMMY_CLIENT_ID,
  // *******************************
  // ================================
  // this section is for build lib, can comment when developing
  // ================================
  _items: [
    {
      title: true,
      name: 'Precise Financial',
      class: '',
      wrapper: {
        element: '',
        attributes: {}
      }
    },
    {
      name: 'Dashboard',
      icon: 'icon-eye',
      to: {name: 'PFDashboard', params: {client_id: DUMMY_CLIENT_ID}},
      badge: {
        variant: 'primary'
      }
    }
  ],
  set setItem (clientID) {
    var self = this
    self._items = [
      {
        title: true,
        name: 'Precise Financial',
        class: '',
        wrapper: {
          element: '',
          attributes: {}
        }
      },
      {
        name: 'Dashboard',
        icon: 'icon-eye',
        to: {name: 'PFDashboard', params: {client_id: clientID}},
        badge: {
          variant: 'primary'
        }
      }
    ]
    self.clientId = clientID
  },
  get items () {
    return this._items
  },
  set addItems (arr) {
    if (arr && arr.length) {
      this._items = concat(arr, this._items)
    }
  }
  // ================================
}
