// plat-import-lib/src/router/_routerConfig.js
import DefaultContainer from '@/containers/DefaultContainer'

const routes = {
  path: '/pf',
  name: 'PFDefaultContainer',
  meta: {
    label: 'Home'
  },
  component: DefaultContainer,
  // redirect: { name: 'Dashboard' },
  children: [
    {
      path: 'step_1',
      meta: {
        label: 'Import',
        reloadPermissions: true,
        title: 'Import'
      },
      name: 'Step1Import',
      component: { template: `<Step1Import></Step1Import>` }
    },
    {
      path: 'step_2',
      meta: {
        label: 'Validate',
        reloadPermissions: true,
        title: 'Validate'
      },
      name: 'Step2Validate',
      component: { template: `<Step2Validate></Step2Validate>` }
    },
    {
      path: 'step_3',
      meta: {
        label: 'Process',
        reloadPermissions: true,
        title: 'Process'
      },
      name: 'Step3Process',
      component: { template: `<Step3Process></Step3Process>` }
    },
    {
      path: 'step_4',
      meta: {
        label: 'Result',
        reloadPermissions: true,
        title: 'Result'
      },
      name: 'Step4Result',
      component: { template: `<Step4Result></Step4Result>` }
    }
  ]
}

export default routes
