// plat-import-lib/src/shared/utils.js
export function random (min, max) {
  return Math.floor(Math.random() * (max - min + 1) + min)
}

/**
 * Randomize array element order in-place.
 * Using Durstenfeld shuffle algorithm.
 */
export const shuffleArray = (array) => {
  for (let i = array.length - 1; i > 0; i--) {
    let j = Math.floor(Math.random() * (i + 1))
    let temp = array[i]
    array[i] = array[j]
    array[j] = temp
  }
  return array
}

export const buildUrlQueryString = (object) => {
  let url = ''
  let keys = Object.keys(object)
  keys.forEach((k, i) => {
    if (object[k] !== '' && object[k] !== undefined && object[k] !== null) {
      url += `${k}=${encodeURIComponent(object[k])}&`
    }
  })
  return `${url.replace(/([&?])$/g, '')}`
}

export const isValidASINOrUPC = (value) => {
  if (!value) {
    return true
  }
  let list = value.replace(/[\s+,]/g, '\n').split('\n').filter(e => !!e)
  if (!list.length) {
    return false
  }
  const regexps = [
    // ASIN regex
    new RegExp(`^[A-Z0-9]{10}$`),
    // UPC regex
    new RegExp(`^[0-9]{12}$`),
    // EAN regex
    new RegExp(`^[0-9]{13}$`)
  ]
  return list
    .map(
      a => regexps
        .map(r => !!a.trim().match(r))
        .filter(r => r === true).length > 0
    )
    .filter(r => r === false).length === 0
}

export const isASIN = (string) => {
  let asinRegex = /^[A-Z0-9]{10}$/
  if (string.match(asinRegex) !== null) {
    return true
  } else {
    return false
  }
}

export const isUPC = (string) => {
  let upcRegex = /^[0-9]{12}$/
  if (string.match(upcRegex) !== null) {
    return true
  } else {
    return false
  }
}

export const isEAN = (string) => {
  let eanRegex = /^[0-9]{13}$/
  if (string.match(eanRegex) !== null) {
    return true
  } else {
    return false
  }
}

export const landingComponents = [
  {
    router_name: 'MTBrandOverview',
    view_permission_code: 'ovview',
    edit_permission_code: 'ovedit'
  },
  {
    router_name: 'MTOperationalAnalytics',
    view_permission_code: 'oaview',
    edit_permission_code: 'oaedit'
  },
  {
    router_name: 'MTProductListings',
    view_permission_code: 'plview',
    edit_permission_code: 'pledit'
  },
  {
    router_name: 'MTMarketingPerformance',
    view_permission_code: 'mpview',
    edit_permission_code: 'mpedit'
  },
  {
    router_name: 'MTCustomerReviews',
    view_permission_code: 'crview',
    edit_permission_code: 'credit'
  },
  {
    router_name: 'MTCategoryMarketShare',
    view_permission_code: 'cmview',
    edit_permission_code: 'cmedit'
  },
  {
    router_name: 'MTSellersAndInvestigations',
    view_permission_code: 'siview',
    edit_permission_code: 'siedit'
  },
  {
    router_name: 'MTAddCustomDashboard',
    view_permission_code: 'cdview',
    edit_permission_code: 'cdedit'
  }
]
